src/components/AddressBook/SetContactDialog.vue

```diff
@@ -0,0 +1,251 @@
+        <s-tooltip :content="copyValueAssetId" placement="top">
+          <p class="wallet-send-address-formatted" @click="handleCopyAddress(formattedSoraAddress, $event)">
+            {{ formattedSoraAddress }}
+          </p>
+        </s-tooltip>
+      </template>
+      <s-input
+        class="set-address__input"
+        :placeholder="t('addressBook.identity')"
+        v-model="onChainIdentity"
+        :disabled="true"
+      />
+      <div class="set-address__btn">
+        <s-button @click="setContact" type="primary" class="s-typography-button--large" :disabled="btnDisabled">
+          {{ btnText }}
+        </s-button>
+      </div>
+    </div>
+  </dialog-base>
+</template>
+
+<script lang="ts">
+import debounce from 'lodash/debounce';
+import { Component, Mixins, Prop, Ref, Watch } from 'vue-property-decorator';
+
+import { state, getter, mutation } from '@/store/decorators';
+import type { AccountBook, Book, PolkadotJsAccount } from '@/types/common';
+import { formatSoraAddress } from '@/util';
+
+import { api } from '../../api';
+import DialogBase from '../DialogBase.vue';
+import CopyAddressMixin from '../mixins/CopyAddressMixin';
+import DialogMixin from '../mixins/DialogMixin';
+import LoadingMixin from '../mixins/LoadingMixin';
+import TranslationMixin from '../mixins/TranslationMixin';
+
+@Component({
+  components: {
+    DialogBase,
+  },
+})
+export default class SetContactDialog extends Mixins(DialogMixin, TranslationMixin, LoadingMixin, CopyAddressMixin) {
+  @state.account.book book!: Book;
+  @state.account.polkadotJsAccounts polkadotJsAccounts!: Array<PolkadotJsAccount>;
+  @getter.account.account account!: PolkadotJsAccount;
+  @mutation.account.setAddressToBook setAddressToBook!: (record: AccountBook) => void;
+
+  @Prop({ default: '', type: String }) readonly prefilledAddress!: string;
+  @Prop({ default: false, type: Boolean }) readonly isEditMode!: boolean;
+
+  @Ref('address') private readonly addressInput!: HTMLInputElement;
+
+  @Watch('isVisible')
+  async handlePrefilledValues(isVisible: boolean): Promise<void> {
+    if (!isVisible) {
+      this.address = '';
+      this.name = '';
+      return;
+    }
+
+    if (this.prefilledAddress) {
+      this.address = this.prefilledAddress;
+      this.name = this.book[this.prefilledAddress];
+    }
+
+    await this.$nextTick();
+    this.addressInput.focus();
+  }
+
+  @Watch('address')
+  handleAddressInput(address: string): void {
+    if (!api.validateAddress(address)) {
+      this.onChainIdentity = this.t('addressBook.none');
+      return;
+    }
+
+    this.defineIdentity(address);
+  }
+
+  address = '';
+  name = '';
+  onChainIdentity = this.t('addressBook.none');
+
+  defineIdentity = debounce(this.getIdentity, 500);
+
+  setContact(): void {
+    this.setAddressToBook({ address: formatSoraAddress(this.address), name: this.name });
+    this.$root.$emit('updateAddressBook');
```

Comment:
do you really need root events here?

Reply:
it's useful as SetContactDialog.vue and AddressBookDialog.vue are separate components, once you set new contact in SetContactDialog.vue (also setting in localStorage), you have to update the current list in AddressBookDialog.vue  and resort it alphabetically.